src/snabbdom.ts

```diff
@@ -1,55 +1,90 @@
-/* global module, document, Node */
-import {Module} from './modules/module';
-import {Hooks} from './hooks';
-import vnode, {VNode, VNodeData, Key} from './vnode';
-import * as is from './is';
-import htmlDomApi, {DOMAPI} from './htmldomapi';
-
-function isUndef(s: any): boolean { return s === undefined; }
-function isDef(s: any): boolean { return s !== undefined; }
+/* global Node */
+import {Hooks, InsertHook, VNode, VNodes, Key, VText, VComment,
+        Module, VElement, VFragment, DOMAPI, PreHook, CreateHook, UpdateHook, DestroyHook, RemoveHook, PostHook} from './types';
 
-type VNodeQueue = Array<VNode>;
+const { isArray } = Array;
 
-const emptyNode = vnode('', {}, [], undefined, undefined);
+const ELEMENT_NODE = 1, TEXT_NODE = 3, COMMENT_NODE = 8, DOCUMENT_FRAGMENT_NODE = 11;
 
-function sameVnode(vnode1: VNode, vnode2: VNode): boolean {
-  return vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
-}
+function isUndef(s: any): s is undefined { return s === undefined; }
+function isDef(s: any): boolean { return s !== undefined; }
 
-function isVnode(vnode: any): vnode is VNode {
-  return vnode.sel !== undefined;
-}
+type VNodeQueue = VNode[];
+
+const emptyNode: VNode = {
+  nt: 0,
+  sel: '',
+  data: {},
+  children: undefined,
+  text: undefined,
+  elm: undefined,
+  key: undefined,
+};
 
 type KeyToIndexMap = {[key: string]: number};
 
 type ArraysOf<T> = {
   [K in keyof T]: (T[K])[];
 }
 
-type ModuleHooks = ArraysOf<Module>;
+export interface PatchFunction {
+  (oldVnode: VNode, vnode: VNode): VNode;
+  children(parentElm: Node, oldCh: VNodes, newCh: VNodes): VNodes;
+}
+
+function defaultCompareFn(vnode1: VNode, vnode2: VNode): boolean {
+  return vnode1.nt === vnode2.nt && vnode1.key === vnode2.key && vnode1.sel === vnode2.sel;
+}
+
+function isVNode(vnode: any): vnode is VNode {
+  return vnode != null && vnode !== undefined;
```

Comment:
If vnode is undefined, the second check (`vnode !== undefined`) will not happen making it redundant. The loose equality in first check will do implicit coercion between null and undefined

Reply:
ah! good catch! will fix it.

Reply:
You may want also to swap the check order to take advantage of short circuit evaluation, given that undefined values will be more frequent than null

Reply:
Hmm, I don't think that's the case. In the places where we call `isVNode()`, usually we encounter a real vnode, or null, undefined is probably less common since `h` normalizes `undefined` to null in children collection: https://github.com/snabbdom/snabbdom/pull/362/files#diff-8fa4f0f1ff07b4217fd7b7626c8cf01eR117